Use glslify-deps and glslify-bundle

pqml opened the pull request:
Latest versions of glslify use glslify-deps and glslify-bundle to create the shader dependency tree and the bundled sourcecode.

The changes update glslify-loader to use these dependencies instead of glslify.

I've also adapted the transform ability from @s9k [pull-request](https://github.com/stackgl/glslify-loader/pull/9) and I've added the ability to use post-transforms like on the current version of glslify.

pqml committed:
Update code to work with glslify-deps and glslify-bundle

pqml committed:
1.0.3

pqml committed:
Add post-transforms ability

pqml committed:
1.0.4

rreusser commented:
@pqml Completely superficial feedback:

- probably leave out the addition of yarn.lock or at the very least move discussion of that point to a separate issue/PR.
- best not to bump the version number in a PR. That can be bumped once it's merged and published.

Beyond that and for anyone seeing this, the best way to move forward is to try out the branch and report back with whether or not it worked and review the code. I don't happen to have the time or need to dig into it at the moment, but I'm glad to do what I can to facilitate the process for anyone who might. Thanks!

Commit:
Rewrite Readme, add tests, add webpack 4 support

Commit:
Leave version number as the current one

pqml commented:
hi @rreusser, I've followed your feedbacks and made some other adjustements: 

- I've rewritten the readme accordingly to Webpack 3 / Webpack 4 syntaxes
- I've added some glslify-loader + webpack tests
  - test the loader both from an inline require statement and a webpack config file
  - test including local & external glsl dependencies
  - test transforms and post transforms (using `glslify-fancy-imports` and `glslify-hex`)

In my office we used this branch during the past 6 months in several small to medium-sized webpack projects and it works fine, even with advanced webpack features like shader hot reloading (glsl sub-dependencies are also watched by webpack)

pqml commented:
I've just seen this PR: https://github.com/glslify/glslify-loader/pull/19

I think the main difference is that my PR uses `glslify-deps` and `glflisy-bundle` instead of relying on `glslify.compile`. There are several advantages to do this: 

  - We can use the dependency tree generated by `glslify-deps` to make webpack watch all glslify sub-dependencies. This is very useful to do livereload / hotreload on the shader files
  - Since `glslify-deps` + `glslisy-bundle` are just a fragment of glslify, they are using less dependencies than `glslify`, thus reducing `glslify-loader` footprint
  - [babel-plugin-glslify](https://github.com/glslify/babel-plugin-glslify/blob/master/lib/exec-glslify.js) (which share the same prupose as glslify-loader) is already using these two dependencies and it seems to work pretty well

kevzettler commented on package.json:
```diff
@@ -2,21 +2,35 @@
+  "main": "glslify-loader.js",
   "license": "MIT",
   "scripts": {
-    "start": "webpack-dev-server --colors",
```
Why was the dev server build removed?

pqml replied:
I rewrote a lot of code and remove the dev server build in the process. 
A new example is available, using webpack 4 and the new `webpack-serve` module. It also shows how hot module reloading can be used to livereload shaders

marcofugaro replied:
[webpack-serve](https://github.com/webpack-contrib/webpack-serve) has been deprecated now...

kevzettler commented:
Tests pass for me locally

```
Kevs-MBP-2:glslify-loader kev$ npm test

> glslify-loader@1.0.2 test /Users/kev/code/glslify-loader
> npm run tape --silent


  simple glsl file with local & external dependencies

    ✔ #define GLSLIFY 1 was added
    ✔ node module glsl-noise was included
    ✔ cnoise was not renamed
    ✔ local dependency shader-export.glsl was included

  glslify-loader using a transform option (glslify-fancy-imports)

    ✔ local dependency shader-export.glsl was included

  glslify-loader using post transforms (glslify-hex)

    ✔ #00FF00 transformed away with glslify-hex

  inline use

    ✔ #define GLSLIFY 1 was added
    ✔ node module glsl-noise was included


  total:     8
  passing:   8
  duration:  4.1s

```

kevzettler commented:
This PR fixes https://github.com/rreusser/glsl-solid-wireframe/issues/3

and doesn't seem to break any of my existing GLSLify pipeline

kevzettler commented:
I have taken a pass and reviewed this PR. The test suite looks good and passes. Using this version of the `glslify-loader` module does not appear to break a current glslify webpack pipline I am using on http://plebland.com/. This PR addresses This PR fixes rreusser/glsl-solid-wireframe#3 

cc @rreusser in regards to discussion at https://github.com/rreusser/glsl-solid-wireframe/issues/3#issuecomment-394447609 

I'm not sure what other criteria can be met here to further establish trust and confidence to get this merged. Scaling open source contribution and collaboration is an exercise in scaling personal trust and risk aversion. Its a matter or extending and relinquishing control to the community. @pqml has put a non-trivial amount of effort into this work and I trust the community would be responsive in addressing any future issues.

rreusser commented:
@kevzettler Apologies if I was too uptight about getting this merged! (And have also been out for the last week; pardon the delay.) If you or @pqml would like to help, I'm more than happy to add you as collaborators so we can merge and publish. I try to help keep the wheels turning, but I'm always careful to make sure I don't step on toes and merge code that may or may not reflect the original author's intentions. Of course that sorta blocks progress. Thanks for pushing me on that.

@pqml Apologies if I slowed things down here. Please don't interpret that as a lack of gratitude for the patch! 👏 

A summary for @hughsk from the original PR:

> Latest versions of glslify use glslify-deps and glslify-bundle to create the shader dependency tree and the bundled sourcecode.
> 
> The changes update glslify-loader to use these dependencies instead of glslify.

Unless you have objections or requests @hughsk, I'll move forward with double-checking, merging and published this PR so that people can use the latest version of glslify (which includes important fixes like knowing that `fwidth` is not a renamable function) with webpack.

pqml committed:
Add an example with hot reloading and regl

pqml commented:
@rreusser Sure, I'll be happy to help maintaining this repo !
I've just added an example using `regl` and `webpack-serve` to simplify your double-checking.
This example also shows how to implement hot module reloading with glslify.

rreusser commented:
@pqml I'm on the wrong computer at the moment so that npm creds are an issue, but I've added you as a collaborator on this repo. Again, more than happy to add you to the glslify org as well. I'll see if I can clone and test this this evening, but if the message is that it's broken now and it works with these modifications, then it sounds like it's the path forward. Thanks!! 🎉

Apidcloud commented:
Sorry, I hadn't seen this PR at the time I created https://github.com/glslify/glslify-loader/pull/19. I think it does in fact go further, particularly with the HMR. I'll give it a try later and let you know if it works in my use case!

pqml commented:
Hi, sorry for the delay

@rreusser did you have some time to test if the changes were ok for you ? You added me as a collaborator but I don't want to merge anything without your approval, plus I wouldn't be able to update the npm package anyway

Apidcloud commented:
As per my use cases, it worked just fine 🙌 

I think we can close #19 

**Edit:**
I've also closed the other related PRs (#6, #9, #16, #17)

Apidcloud commented:
Can we merge this and maybe (major) release? I think that's the best approach to check if people have issues or not.

marcofugaro commented:
I think so @Apidcloud, we can do other PRs if we find other issues later. For example the fact that `webpack-serve` has been deprecated. I can do another PR to update it later.

Apidcloud commented:
Yeah, I think it's the best option. @rreusser can you make the release?

pqml commented:
Hi, I can merge the PR but this is pointless as I don't have any rights to publish a new major version on npm.

`webpack-serve` is only used as devDependency in the example but I will switch to `webpack-dev-server` when I have some free time

Apidcloud commented:
Yeah, same. Let's wait for @rreusser

rreusser commented:
I'm glad to hear this is working for people, and sounds the right move is to merge, bump to v2.0.0, and publish. Thanks so much for sticking with this!

Okay, I've merged and published as v2.0.0. 🎉 

Would anyone like me to add you as an npm collaborator? (*)

(*) in light of the event-stream fiasco, not without consideration and a once-over for malware. You all have stuck with this and I appreciate it!! :clap:😄 One final apology if I've slowed things down.
